File: packages/core/src/fourPercent.ts
import type { FireNumber, WithdrawalAnalysis } from "./types.js";

/** Default safe withdrawal rate (Bengen 1994 / Trinity study). */
export const DEFAULT_WITHDRAWAL_RATE = 0.04;

/**
 * Calculates how much you can safely withdraw each year (and month) from a
 * given portfolio using the safe withdrawal rate rule.
 *
 * @param portfolioValue   Total portfolio value in dollars.
 * @param withdrawalRate   Annual withdrawal rate as a decimal (default 0.04).
 */
export function safeWithdrawal(
  portfolioValue: number,
  withdrawalRate: number = DEFAULT_WITHDRAWAL_RATE
): WithdrawalAnalysis {
  const annualWithdrawal = portfolioValue * withdrawalRate;
  return {
    portfolioValue,
    annualWithdrawal,
    monthlyWithdrawal: annualWithdrawal / 12,
    withdrawalRate,
  };
}

/**
 * Calculates the portfolio size (FIRE number) needed to sustainably cover a
 * given annual expense level — the inverse of the withdrawal rate rule.
 *
 * Rule of thumb: multiply annual expenses by 25 (at the 4% rate).
 *
 * @param annualExpenses   Desired annual spending in dollars.
 * @param withdrawalRate   Annual withdrawal rate as a decimal (default 0.04).
 */
export function fireNumber(
  annualExpenses: number,
  withdrawalRate: number = DEFAULT_WITHDRAWAL_RATE
): FireNumber {
  return {
    annualExpenses,
    requiredPortfolio: annualExpenses / withdrawalRate,
    withdrawalRate,
  };
}

/**
 * Returns the percentage of the FIRE number already achieved.
 *
 * @param currentPortfolio   Current portfolio value.
 * @param targetPortfolio    Required portfolio (from fireNumber).
 */
export function fireProgress(
  currentPortfolio: number,
  targetPortfolio: number
): number {
  if (targetPortfolio <= 0) return 0;
  return Math.min(currentPortfolio / targetPortfolio, 1);
}

/**
 * Estimates how many years of expenses the current portfolio can sustain at a
 * given withdrawal rate, assuming zero growth (conservative floor).
 */
export function yearsOfExpenses(
  portfolioValue: number,
  annualExpenses: number
): number {
  if (annualExpenses <= 0) return Infinity;
  return portfolioValue / annualExpenses;
}

File: apps/ios/RetirementPlanner/Core/FourPercent.swift
// 4% rule / FIRE number utilities — mirrors packages/core/src/fourPercent.ts

/// Namespace for the 4% safe-withdrawal-rate rule constant (Bengen 1994 / Trinity study).
enum FourPercentRule {
    static let defaultWithdrawalRate: Double = 0.04
}

/// Annual and monthly safe withdrawal amounts for a given portfolio.
func safeWithdrawal(
    portfolioValue: Double,
    withdrawalRate: Double = FourPercentRule.defaultWithdrawalRate
) -> WithdrawalAnalysis {
    let annual = portfolioValue * withdrawalRate
    return WithdrawalAnalysis(
        portfolioValue: portfolioValue,
        annualWithdrawal: annual,
        monthlyWithdrawal: annual / 12,
        withdrawalRate: withdrawalRate
    )
}

/// Portfolio size (FIRE number) needed to sustainably fund a given annual expense level.
/// Rule of thumb: multiply annual expenses by 25 at the 4% rate.
func fireNumber(
    annualExpenses: Double,
    withdrawalRate: Double = FourPercentRule.defaultWithdrawalRate
) -> FireNumber {
    FireNumber(
        annualExpenses: annualExpenses,
        requiredPortfolio: annualExpenses / withdrawalRate,
        withdrawalRate: withdrawalRate
    )
}

/// Fraction of the FIRE number already achieved, clamped to [0, 1].
func fireProgress(currentPortfolio: Double, targetPortfolio: Double) -> Double {
    guard targetPortfolio > 0 else { return 0 }
    return min(currentPortfolio / targetPortfolio, 1)
}

/// How many years of expenses the portfolio can cover at zero growth (conservative floor).
func yearsOfExpenses(portfolioValue: Double, annualExpenses: Double) -> Double {
    guard annualExpenses > 0 else { return .infinity }
    return portfolioValue / annualExpenses
}
